src/sparsearraydok.jl

```diff
@@ -1,14 +1,56 @@
 using Accessors: @set
-using Dictionaries: Dictionary, IndexError, set!
+using Dictionaries: AbstractDictionary, Dictionary, IndexError, set!
 
 function default_getunstoredindex(a::AbstractArray, I::Int...)
   return zero(eltype(a))
 end
 
+const DOKStorage{T,N} = Dictionary{CartesianIndex{N},T}
+
 struct SparseArrayDOK{T,N,F} <: AbstractSparseArray{T,N}
-  storage::Dictionary{CartesianIndex{N},T}
+  storage::DOKStorage{T,N}
   size::NTuple{N,Int}
   getunstoredindex::F
+
+  # bare constructor
+  function SparseArrayDOK{T,N,F}(
+    ::UndefInitializer, size::Dims{N}, getunstoredindex::F
+  ) where {T,N,F}
+    storage = DOKStorage{T,N}()
+    return new{T,N,F}(storage, size, getunstoredindex)
+  end
+
+  # unchecked constructor from data
+  function SparseArrayDOK{T,N,F}(
+    storage::DOKStorage{T,N}, size::Dims{N}, getunstoredindex::F
+  ) where {T,N,F}
+    return new{T,N,F}(storage, size, getunstoredindex)
+  end
+end
+
+# undef constructors
+function SparseArrayDOK{T}(
+  ::UndefInitializer, dims::Dims, getunstoredindex=default_getunstoredindex
+) where {T}
+  all(≥(0), dims) || throw(ArgumentError("Invalid dimensions: $dims"))
+  N = length(dims)
+  F = typeof(getunstoredindex)
+  return SparseArrayDOK{T,N,F}(undef, dims, getunstoredindex)
+end
+function SparseArrayDOK{T}(::UndefInitializer, dims::Int...) where {T}
+  return SparseArrayDOK{T}(undef, dims)
+end
+
+# checked constructor from data: use `setindex!` to validate input
+# does not take ownership of `storage`!
+function SparseArrayDOK(
+  storage::Union{AbstractDictionary{I,T},AbstractDict{I,T}}, dims::Dims{N}, unstored...
+) where {N,I<:Union{Int,CartesianIndex{N}},T}
+  A = SparseArrayDOK{T}(undef, dims, unstored...)
+  for (i, v) in pairs(storage)
+    A[i] = v
```

Comment:
Does this work for `AbstractDictionary`? Doesn't it need to be `set!` or `insert!`?

Reply:
Note that A is not a dictionary, it's the array itself. This is a call to setindex, which would translate that into the correct dictionary call

Reply:
If so, probably we should define a function:
```julia
set_dict!(d::AbstractDictionary, key, value) = set!(d, key, value)
set_dict!(d::AbstractDict, key, value) = (d[key] = value)
```
and use that in code we want to be generic between the two.

Relatedly, I've want to make my own `AbstractDictionary` subtype that defines `setindex!` as `set!`, say called `ExpandableDictionary`, to make it easier to write generic code between `Dict` and `Dictionary`, so then one could write:
```julia
d′ = expandable(d)
d′[key] = value
```
which could be generic between `Dict` and `Dictionary`, say if `expandable(d::AbstractDict) = d` and `expandable(d::AbstractDictionary) = ExpandableDictionary(d)`.

Reply:
> Note that A is not a dictionary, it's the array itself. This is a call to setindex, which would translate that into the correct dictionary call

Oh I see, thanks for the clarification!

Reply:
I think this is a bit outside of the scope of this PR in any case :)

The main point is that I'm using `setindex!`, which we can assume already has the correct call to `set!`. Additionally, this ensures that all provided indices are inbounds, because of the boundschecks that (normally) happen with `setindex!`

Reply:
Agreed! Definitely a tangent.